feat(core): Introduce workflowPublishHistory table (no-changelog)

CharlieKolb opened the pull request:
## Summary

Introduce a new table which tracks historic workflow activations, relating acting userId where available.

Included some util functions on the repo, we may want to expose the one returning all published versions of a workflow via a new controller at some point. 

Open question is if and how we need to clean this table up over time. 

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/ADO-4295

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

CharlieKolb committed:
Add workflow publish history

CharlieKolb committed:
fix tests, deactivate if stay inactive

CharlieKolb committed:
mark as nullable

CharlieKolb committed:
nullable user

CharlieKolb committed:
exclude user

CharlieKolb committed:
lint

CharlieKolb committed:
push userId instead

CharlieKolb committed:
try without versionId foreign key

CharlieKolb committed:
readd versionId, use uuid for user

CharlieKolb committed:
fix migration and repo

CharlieKolb committed:
use user instead

CharlieKolb committed:
Merge remote-tracking branch 'origin/master' into ADO-4295_3

CharlieKolb committed:
fix new tests

CharlieKolb committed:
lowercase column name

CharlieKolb committed:
fix migration

CharlieKolb committed:
fix user relation

CharlieKolb committed:
fix multi main publish history tracking

dariacodes commented on packages/@n8n/db/src/repositories/workflow-publish-history.repository.ts:
```diff
@@ -0,0 +1,63 @@
+
+	async getPublishedVersions(workflowId: string, includeUser?: boolean) {
+		const select: Array<keyof WorkflowPublishHistory> = ['versionId', 'createdAt'];
+		let relations = {};
```
why not make it an array and use `const`? anyway for all the relations you'll have to send `true`

dariacodes commented on packages/cli/src/active-workflow-manager.ts:
```diff
@@ -918,13 +931,33 @@ export class ActiveWorkflowManager {
+
+		await this.workflowPublishHistoryRepository.addRecord({
+			workflowId,
+			versionId: null,
```
don't we need to know which version was deactivated? this field should never be null, right?

dariacodes commented on packages/cli/src/environments.ee/source-control/source-control-import.service.ee.ts:
```diff
@@ -735,19 +735,26 @@ export class SourceControlImportService {
+			await this.activeWorkflowManager.remove(
+				existingWorkflow.id,
+				userId,
+				importedWorkflow.activeVersionId ? 'update' : 'deactivate',
```
These values are so confusing (in our repo I mean 😆 )! But yeah, I guess with the current logic we should always send "update" if it currently has the active version

dariacodes reviewed (commented):
Nice work! I left a few comments

CharlieKolb committed:
pr feedback 1

RicardoE105 commented on packages/@n8n/db/src/migrations/common/1763387043735-CreateWorkflowPublishHistoryTable.ts:
```diff
@@ -0,0 +1,54 @@
+				column('id').int.primary.autoGenerate2,
+				column('workflowId').varchar(36).notNull,
+				column('versionId').varchar(36).notNull,
+				column('status').varchar(36).notNull,
```
can we add a .comment with the possible values?

RicardoE105 committed:
Fix version id column

 versionId was defined as NOT NULL but the foreign key constraint specified onDelete: 'SET NULL'. When a row in workflow_history is deleted

RicardoE105 committed:
Add comments to the columns

RicardoE105 committed:
make entity match the tables in the db

RicardoE105 committed:
Add index in createdAt column

RicardoE105 committed:
add tests for the workflow-publish-history

CharlieKolb committed:
cascade versionId on version deletion

CharlieKolb commented on packages/@n8n/db/src/entities/workflow-publish-history.ts:
```diff
@@ -0,0 +1,38 @@
+import { User } from './user';
+
+@Entity()
+@Index(['workflowId', 'versionId'])
```
Reading into it a bit I think including `createdAt` makes sense, but status or mode doesn't, since we'll get the ordering with the createdAt index, and we won't ever have a meaningful amount of `deactivated` entries before the first `activated` one, so we don't mind looping through those until we find the latest one. lmk what you think

CharlieKolb committed:
sort tables

CharlieKolb committed:
fix tests, return last activated rather than last activate mode

tomi commented on packages/@n8n/db/src/entities/types-db.ts:
```diff
@@ -416,3 +416,5 @@ export type WorkflowHistoryUpdate = Omit<
 	'versionId' | 'workflowId' | 'createdAt' | 'updatedAt'
 >;
+
+export type WorkflowPublishHistoryMode = 'activate' | 'update' | 'deactivate';
```
What is `update`? Can we add a JSDoc what these mean?

tomi commented on packages/@n8n/db/src/migrations/common/1763387043735-CreateWorkflowPublishHistoryTable.ts:
```diff
@@ -0,0 +1,64 @@
+				column('userId').uuid,
+			)
+			.withCreatedAt.withIndexOn('workflowId')
+			.withIndexOn(['workflowId', 'versionId'])
```
This is way too many indexes. Please think what the query patterns are and only add indexes to support those

CharlieKolb replied:
I agree! I was gonna go with a `workflowId` index for joins and a workflowId, versionId access to quickly find the latest activation time for a given version or all versions of a given workflow.

Does that sound reasonable or should we only add one by workflowId?

tomi commented on packages/@n8n/db/src/migrations/common/1763387043735-CreateWorkflowPublishHistoryTable.ts:
```diff
@@ -0,0 +1,64 @@
+				tableName: 'user',
+				columnName: 'id',
+				onDelete: 'SET NULL',
+			});
```
Would be nice to add a `.withEnumCheck` constraint for the enum columns

tomi reviewed (commented):
Only reviewed the entity and migration

dariacodes commented on packages/@n8n/db/src/migrations/common/1763387043735-CreateWorkflowPublishHistoryTable.ts:
```diff
@@ -0,0 +1,64 @@
+			.withColumns(
+				column('id').int.primary.autoGenerate2,
+				column('workflowId').varchar(36).notNull,
+				column('versionId').varchar(36),
```
We lost `notNull` here

dariacodes commented on packages/@n8n/db/src/repositories/workflow-publish-history.repository.ts:
```diff
@@ -0,0 +1,58 @@
+		});
+	}
+
+	async getLastActivatedVersion(workflowId: string) {
```
It might be that we don't actually need this method right now. The initial thinking was that `activateAll` would activate the last activated version, but now that we're removing it completely, we might not need this anymore

dariacodes commented on packages/cli/test/integration/database/repositories/workflow-publish-history.repository.test.ts:
```diff
@@ -0,0 +1,438 @@
+import { createUser } from '../../shared/db/users';
+
+const id1 = '5ef472d2-9253-452c-b0fe-8eb78fb3c43b';
+const id2 = 'bf36ce2c-baf6-4b51-9f01-065c76d5cb0c';
```
We usually use `uuid` for this:
```import { v4 as uuid } from 'uuid';)```

CharlieKolb committed:
Tomi Feedback

CharlieKolb committed:
fix tests

dariacodes commented on packages/cli/src/active-workflow-manager.ts:
```diff
@@ -949,6 +978,24 @@ export class ActiveWorkflowManager {
 		}
 	}
 
+	async trackRemovalInPublishingHistory(
```
Yeah, the repo shouldn't be fetching the workflow, that's why I said "service".
Alright, I guess it's fine to keep it here. It is related to active workflows anyway 😄

CharlieKolb committed:
Move tracking out of active workflow manager

CharlieKolb committed:
update migration timestamp, include updatedAt

CharlieKolb committed:
lint

CharlieKolb committed:
unflake the test

CharlieKolb committed:
escape table name

CharlieKolb committed:
fix migration

CharlieKolb committed:
actually escape the table name

CharlieKolb committed:
more escaping

CharlieKolb committed:
Remove extra repo functions and clean up PR

CharlieKolb committed:
fix test

CharlieKolb committed:
use uuid

CharlieKolb committed:
unflake tests

dariacodes commented on packages/cli/src/environments.ee/source-control/source-control-import.service.ee.ts:
```diff
@@ -749,6 +759,13 @@ export class SourceControlImportService {
+				versionId: importedWorkflow.activeVersionId ?? existingWorkflow.activeVersionId,
+				status: importedWorkflow.activeVersionId ? 'activated' : 'deactivated',
+				userId,
+			});
```
could you make two calls instead?

CharlieKolb committed:
PR feedback

CharlieKolb committed:
add comment

CharlieKolb committed:
add controller tests

CharlieKolb committed:
lint

CharlieKolb committed:
add to import service, remove from public api

CharlieKolb committed:
Merge remote-tracking branch 'origin/master' into ADO-4295_3

CharlieKolb committed:
lint

CharlieKolb committed:
fix tests and pr feedback

CharlieKolb committed:
add null check

CharlieKolb committed:
lint

CharlieKolb committed:
use activeVersionId in helper function

CharlieKolb committed:
add deactived case and amend test cases

dariacodes reviewed (dismissed):
Looks good 🚀

CharlieKolb committed:
fix deactivate in sc import

tomi commented on packages/cli/src/workflows/workflow.service.ee.ts:
```diff
@@ -329,8 +331,7 @@ export class EnterpriseWorkflowService {
 		// 6. deactivate workflow if necessary
-		const wasActive = workflow.activeVersionId !== null;
-		if (wasActive) {
+		if (workflow.activeVersionId !== null) {
```
Why are we reverting these? I don't think that's needed for this change? Also IMHO the previous one was clearer

tomi commented on packages/cli/src/workflows/workflow.service.ts:
```diff
@@ -460,6 +462,15 @@ export class WorkflowService {
 					publicApi,
 				});
+				if (workflow.activeVersionId !== null) {
+					// sanity check - activeVersionId should always be defined at this stage
```
Better to use `assert` if this is an invariant

tomi reviewed (commented):
DB changes look good. Couple questions and comments

CharlieKolb commented on packages/@n8n/db/src/entities/workflow-publish-history.ts:
```diff
@@ -0,0 +1,35 @@
+	mode: WorkflowPublishHistoryMode | null;
+
+	@Column({ type: 'varchar', nullable: true })
+	userId: string | null;
```
It's nullable because we don't always have an associated user when activating a workflow, e.g. from cli or (previously) from activateAll().

Referencing the actual user does seem preferred to me - we may want to associate profile images or similar and `Firstname Lastname` isn't guaranteed to be unique.
I agree the UX here isn't great with deleted users, but I'd argue the root cause is that we hard delete users rather than deactivating them while keeping a reference in the DB 😅

CharlieKolb commented on packages/@n8n/db/src/migrations/common/1764167920585-CreateWorkflowPublishHistoryTable.ts:
```diff
@@ -0,0 +1,58 @@
+
+		await runQuery(
+			`INSERT INTO ${escapedWphTableName} (${workflowId}, ${versionId}, ${status}, ${createdAt})
+				SELECT we.${id}, we.${activeVersionId}, 'activated', we.${updatedAt}
```
This is best effort indeed - I do think there's some value in indicating whether an active workflow was last modified (and thus "published" in our new world), even if this is potentially off a bit.
Knowing a workflow was changed six months ago vs a week ago will be more valuable than showing the timestamp of the migration, especially since that datetime will be completely unrelated to the given workflow.

CharlieKolb committed:
PR Feedback

CharlieKolb committed:
rename status column to event

tomi reviewed (approved):
Thank you for addressing the feedback 🫶
